File: archive/leetcode/lc-2345.py
from typing import List


# Actually not bad tho not the best... do it faster! Debug faster!
class Solution:
    def visibleMountains(self, peaks: List[List[int]]) -> int:
        peaks = [(y, x) for x, y in peaks]
        peaks.sort()
        vp = [peaks[-1]]
        dups = set()

        for y, x in peaks[:-1][::-1]:
            for my, mx in vp:
                dy = my - y
                dx = abs(mx - x)
                if y != my and dy >= dx:
                    break
                elif y == my:
                    if x == mx:
                        dups.add((y, x))
                        break
                    else:
                        vp.append((y, x))
                        break
            else:
                vp.append((y, x))
        return len(vp) - len(dups)


s = Solution()
try:
    r = s.visibleMountains([[2, 2], [2, 2], [3, 1]])
    assert r == 0
    r = s.visibleMountains([[2, 2], [6, 3], [5, 4]])
    assert r == 2
    r = s.visibleMountains([[1, 3], [1, 3]])
    assert r == 0
    r = s.visibleMountains(
        [
            [12, 10],
            [37, 25],
            [8, 12],
            [8, 36],
            [39, 4],
            [22, 3],
            [36, 19],
            [3, 17],
            [10, 19],
            [19, 38],
            [6, 36],
            [27, 23],
            [4, 29],
            [36, 27],
            [21, 28],
            [9, 11],
            [24, 1],
            [16, 17],
            [16, 9],
            [22, 23],
            [37, 31],
            [34, 17],
            [19, 2],
            [33, 3],
            [12, 14],
            [23, 7],
            [20, 36],
            [7, 36],
            [16, 7],
            [24, 38],
        ]
    )
    assert r == 6
except AssertionError:
    print(r)

File: archive/leetcode/lc-2345.rs
use std::collections::HashSet;

impl Solution {
    // Conclusion: bad heuristic (if condition) can make or break things
    #[allow(clippy::comparison_chain)]
    pub fn visible_mountains(peaks: Vec<Vec<i32>>) -> i32 {
        let mut _peaks = Vec::with_capacity(peaks.len());
        let mut vp = Vec::with_capacity(peaks.len());
        for peak in peaks {
            if let [x, y] = peak[..] {
                _peaks.push((y, x));
            }
        }
        _peaks.sort();
        let mut dups = HashSet::new();
        vp.push(*(_peaks.last().unwrap()));
        'peak: for (y, x) in _peaks[..(_peaks.len() - 1)].iter().rev() {
            for &(my, mx) in vp.iter() {
                let dx = (x - mx).abs();
                let dy = my - y;
                if *y != my && dy >= dx {
                    continue 'peak;
                } else if *y == my {
                    if *x == mx {
                        dups.insert((*y, *x));
                        continue 'peak;
                    } else {
                        vp.push((*y, *x));
                        continue 'peak;
                    }
                }
            }
            vp.push((*y, *x));
        }
        (vp.len() - dups.len()) as i32
    }
}

struct Solution;

fn main() {
    assert_eq!(
        Solution::visible_mountains(vec![vec![2, 2], vec![2, 2], vec![3, 1]]),
        0
    );
    assert_eq!(
        Solution::visible_mountains(vec![vec![2, 2], vec![6, 3], vec![5, 4]]),
        2
    );
    assert_eq!(Solution::visible_mountains(vec![vec![1, 3], vec![1, 3]]), 0);
    assert_eq!(
        Solution::visible_mountains(vec![
            vec![12, 10],
            vec![37, 25],
            vec![8, 12],
            vec![8, 36],
            vec![39, 4],
            vec![22, 3],
            vec![36, 19],
            vec![3, 17],
            vec![10, 19],
            vec![19, 38],
            vec![6, 36],
            vec![27, 23],
            vec![4, 29],
            vec![36, 27],
            vec![21, 28],
            vec![9, 11],
            vec![24, 1],
            vec![16, 17],
            vec![16, 9],
            vec![22, 23],
            vec![37, 31],
            vec![34, 17],
            vec![19, 2],
            vec![33, 3],
            vec![12, 14],
            vec![23, 7],
            vec![20, 36],
            vec![7, 36],
            vec![16, 7],
            vec![24, 38]
        ]),
        6
    )
}
